### ISSUE TITLE
[Bug]: Azure OpenAI sometimes sends responses lacking a delta or choice key in the response dict - this causes a silent failure.

### ISSUE THREAD

**Issue by MikeRenwick-ICG** [run-llama/llama_index#8995]
### Bug Description

Azure OpenAI (for example GPT-4-32k-0613) sometimes returns packets that are missing choices & delta keys.  This causes a silent failure.

### Version

llama-index-0.9.3.post1

### Steps to Reproduce

1) Setup an [Azure OpenAI deployment and llamaindex example](https://docs.llamaindex.ai/en/stable/examples/llm/azure_openai.html)
2) Configure [content filtering](https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/content-filter?tabs=warning%2Cpython

Run a query and you'll note that it fails silently.

**Comment by MikeRenwick-ICG** [comment]
I have a fix incoming for this.

**Comment by saveriyo** [comment]
@MikeRenwick-ICG Feel free to ditch/close this PR, but I found this small change worked for me by testing locally with AzureOpenAI and OpenAIAgent.

**Comment by saveriyo** [comment]
@logan-markewich I think unfortunately this doesn't account for the cases @MikeRenwick-ICG mentioned here: https://github.com/run-llama/llama_index/pull/9040#issuecomment-1820533632

It appears that at least OpenAI._stream_chat also faces the same issue handling empty deltas from AzureOpenAI. From what I can tell, this may have to do with the way content filtering is handled by Azure OpenAI vs OpenAI API directly, causing Azure to send a response without a delta.

**Comment by logan-markewich** [comment]
@saveriyo the latest versions of llama-index shoooould be handling this now though? We check explicitly for this

**Comment by saveriyo** [comment]
I'm trying out on llama-index version 0.9.6.post1, and still seeing an issue with AzureOpenAI compared 1 to 1 with OpenAI.

**TLDR: Here's the first chunk from AzureOpenAI that may be breaking function calls:
`ChatCompletionChunk(id='', choices=[], created=0, model='', object='', system_fingerprint=None, prompt_filter_results=[{'prompt_index': 0, 'content_filter_results': {'hate': {'filtered': False, 'severity': 'safe'}, 'self_harm': {'filtered': False, 'severity': 'safe'}, 'sexual': {'filtered': False, 'severity': 'safe'}, 'violence': {'filtered': False, 'severity': 'safe'}}}])`**

Here is what I hope is a minimally reproducible example (just need to provide your API keys in a .env or as env variables):

```python
import os
import asyncio
import logging
from dotenv import load_dotenv
from llama_index.agent import OpenAIAgent
from llama_index.llms import OpenAI, AzureOpenAI
from llama_index.tools import FunctionTool
load_dotenv()

logging.basicConfig(level=logging.INFO)

## Variables
# provider = "open_ai"
# model = "gpt-3.5-turbo-1106"
provider = "azure"
model = "gpt-35-turbo"

api_version, engine = None, None

if provider == "open_ai":
  api_base = "https://api.openai.com/v1"
  api_key = os.getenv("OPENAI_API_KEY")
elif provider == "azure":
  api_base = os.getenv("AZURE_API_BASE")
  api_key = os.getenv("AZURE_API_KEY")
  api_version = os.getenv("AZURE_API_VERSION")
  engine = model

## Inputs
input = "What is the weather in Boston, MA?"
# input = "Hello, how are you?"
system_message = "You must only talk like a pirate!"

## Functions
def get_current_weather(location: str, units: str = "F"):
  """Get the current weather in a given location

  Parameters
  ----------
  location : str
      The city and state, e.g. San Francisco, CA

  units : str
      The units to use (optional). [C or F]

  Returns
  -------
  str
      a sentence indicating the weather
  """
  print("FUNCTION: Getting the current weather!")
  if location == "Boston, MA":
    if units == "C":
      return "12C"
    elif units == "F":
      return "23F"
    else:
      return "Invalid units"
  else:
    return "Invalid location"
  
functions = [FunctionTool.from_defaults(get_current_weather)]

## LLMs
if provider == "open_ai":
  llm = OpenAI(
      model=model,
      temperature=0.0,
      api_base=api_base,
      api_key=api_key,
      api_version=api_version
  )
elif provider == "azure":
  llm = AzureOpenAI(
      engine=engine,
      model=model,
      azure_endpoint=api_base,
      api_version=api_version,
      api_key=api_key,
      temperature=0.0,
  )

## Agent
agent = OpenAIAgent.from_tools(
  tools=functions, 
  llm=llm, 
  system_prompt=system_message,
  verbose=True
  )

## Streaming Async Output
async def stream_output():
  response = await agent.astream_chat(input)
  stream = response.async_response_gen()

  async for message in stream:
    print(message, end="", flush=True)

asyncio.run(stream_output())
```

I added in printing of the `response` within `OpenAI._astream_chat` to see what might be going on:
```python
    async def _astream_chat(
        self, messages: Sequence[ChatMessage], **kwargs: Any
    ) -> ChatResponseAsyncGen:
        message_dicts = to_openai_message_dicts(messages)

        async def gen() -> ChatResponseAsyncGen:
            content = ""
            tool_calls: List[ChoiceDeltaToolCall] = []

            is_function = False
            async for response in await self._aclient.chat.completions.create(
                messages=message_dicts,
                stream=True,
                **self._get_model_kwargs(**kwargs),
            ):
                response = cast(ChatCompletionChunk, response)
                print(f"\nResponse OpenAI: {repr(response)}\n")
                ...
```

Here's with OpenAI:
```
STARTING TURN 1
---------------

INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"

Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGLCRqqJybpPZmwsliDnd7uKmnD', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role='assistant', tool_calls=[ChoiceDeltaToolCall(index=0, id='call_6AvFNzL3m8VOvZXFxn2Pcfte', function=ChoiceDeltaToolCallFunction(arguments='', name='get_current_weather'), type='function')]), finish_reason=None, index=0)], created=1700681601, model='gpt-3.5-turbo-1106', object='chat.completion.chunk', system_fingerprint='fp_eeff13170a')


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGLCRqqJybpPZmwsliDnd7uKmnD', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role=None, tool_calls=[ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='{"', name=None), type=None)]), finish_reason=None, index=0)], created=1700681601, model='gpt-3.5-turbo-1106', object='chat.completion.chunk', system_fingerprint='fp_eeff13170a')


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGLCRqqJybpPZmwsliDnd7uKmnD', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role=None, tool_calls=[ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='location', name=None), type=None)]), finish_reason=None, index=0)], created=1700681601, model='gpt-3.5-turbo-1106', object='chat.completion.chunk', system_fingerprint='fp_eeff13170a')


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGLCRqqJybpPZmwsliDnd7uKmnD', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role=None, tool_calls=[ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='":"', name=None), type=None)]), finish_reason=None, index=0)], created=1700681601, model='gpt-3.5-turbo-1106', object='chat.completion.chunk', system_fingerprint='fp_eeff13170a')


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGLCRqqJybpPZmwsliDnd7uKmnD', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role=None, tool_calls=[ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='Boston', name=None), type=None)]), finish_reason=None, index=0)], created=1700681601, model='gpt-3.5-turbo-1106', object='chat.completion.chunk', system_fingerprint='fp_eeff13170a')


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGLCRqqJybpPZmwsliDnd7uKmnD', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role=None, tool_calls=[ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments=',', name=None), type=None)]), finish_reason=None, index=0)], created=1700681601, model='gpt-3.5-turbo-1106', object='chat.completion.chunk', system_fingerprint='fp_eeff13170a')


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGLCRqqJybpPZmwsliDnd7uKmnD', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role=None, tool_calls=[ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments=' MA', name=None), type=None)]), finish_reason=None, index=0)], created=1700681601, model='gpt-3.5-turbo-1106', object='chat.completion.chunk', system_fingerprint='fp_eeff13170a')


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGLCRqqJybpPZmwsliDnd7uKmnD', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role=None, tool_calls=[ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='"}', name=None), type=None)]), finish_reason=None, index=0)], created=1700681601, model='gpt-3.5-turbo-1106', object='chat.completion.chunk', system_fingerprint='fp_eeff13170a')


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGLCRqqJybpPZmwsliDnd7uKmnD', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role=None, tool_calls=None), finish_reason='tool_calls', index=0)], created=1700681601, model='gpt-3.5-turbo-1106', object='chat.completion.chunk', system_fingerprint='fp_eeff13170a')

=== Calling Function ===
Calling function: get_current_weather with args: {"location":"Boston, MA"}
FUNCTION: Getting the current weather!
Got output: 23F
========================

STARTING TURN 2
---------------
...
```


Here's with AzureOpenAI:
```
STARTING TURN 1
---------------

INFO:httpx:HTTP Request: POST https://iyo.openai.azure.com//openai/deployments/gpt-35-turbo/chat/completions?api-version=2023-10-01-preview "HTTP/1.1 200 OK"

Response OpenAI: ChatCompletionChunk(id='', choices=[], created=0, model='', object='', system_fingerprint=None, prompt_filter_results=[{'prompt_index': 0, 'content_filter_results': {'hate': {'filtered': False, 'severity': 'safe'}, 'self_harm': {'filtered': False, 'severity': 'safe'}, 'sexual': {'filtered': False, 'severity': 'safe'}, 'violence': {'filtered': False, 'severity': 'safe'}}}])


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGdrSzoX9rsmtwBzCOp1Re62xE1', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role='assistant', tool_calls=[ChoiceDeltaToolCall(index=0, id='call_KIyW5cX46hnFnz9e02Z69XCz', function=ChoiceDeltaToolCallFunction(arguments='', name='get_current_weather'), type='function')]), finish_reason=None, index=0, content_filter_results={})], created=1700681619, model='gpt-35-turbo', object='chat.completion.chunk', system_fingerprint='fp_50a4261de5')


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGdrSzoX9rsmtwBzCOp1Re62xE1', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role=None, tool_calls=[ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='{"', name=None), type=None)]), finish_reason=None, index=0, content_filter_results={})], created=1700681619, model='gpt-35-turbo', object='chat.completion.chunk', system_fingerprint='fp_50a4261de5')


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGdrSzoX9rsmtwBzCOp1Re62xE1', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role=None, tool_calls=[ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='location', name=None), type=None)]), finish_reason=None, index=0, content_filter_results={})], created=1700681619, model='gpt-35-turbo', object='chat.completion.chunk', system_fingerprint='fp_50a4261de5')


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGdrSzoX9rsmtwBzCOp1Re62xE1', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role=None, tool_calls=[ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='":"', name=None), type=None)]), finish_reason=None, index=0, content_filter_results={})], created=1700681619, model='gpt-35-turbo', object='chat.completion.chunk', system_fingerprint='fp_50a4261de5')


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGdrSzoX9rsmtwBzCOp1Re62xE1', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role=None, tool_calls=[ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='Boston', name=None), type=None)]), finish_reason=None, index=0, content_filter_results={})], created=1700681619, model='gpt-35-turbo', object='chat.completion.chunk', system_fingerprint='fp_50a4261de5')


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGdrSzoX9rsmtwBzCOp1Re62xE1', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role=None, tool_calls=[ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments=',', name=None), type=None)]), finish_reason=None, index=0, content_filter_results={})], created=1700681619, model='gpt-35-turbo', object='chat.completion.chunk', system_fingerprint='fp_50a4261de5')


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGdrSzoX9rsmtwBzCOp1Re62xE1', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role=None, tool_calls=[ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments=' MA', name=None), type=None)]), finish_reason=None, index=0, content_filter_results={})], created=1700681619, model='gpt-35-turbo', object='chat.completion.chunk', system_fingerprint='fp_50a4261de5')


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGdrSzoX9rsmtwBzCOp1Re62xE1', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role=None, tool_calls=[ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='"}', name=None), type=None)]), finish_reason=None, index=0, content_filter_results={})], created=1700681619, model='gpt-35-turbo', object='chat.completion.chunk', system_fingerprint='fp_50a4261de5')


Response OpenAI: ChatCompletionChunk(id='chatcmpl-8NnGdrSzoX9rsmtwBzCOp1Re62xE1', choices=[Choice(delta=ChoiceDelta(content=None, function_call=None, role=None, tool_calls=None), finish_reason='tool_calls', index=0, content_filter_results={})], created=1700681619, model='gpt-35-turbo', object='chat.completion.chunk', system_fingerprint='fp_50a4261de5')
```

The key difference seems to be the first message from Azure, which has an empty choices field.

**Comment by logan-markewich** [comment]
@saveriyo I encourage you to make a PR if you can then, since you seem to have this figured out.

**Comment by saveriyo** [comment]
The only solution I have found to work is this commit:

https://github.com/run-llama/llama_index/pull/9040/commits/38622b775401d51c6d9907eb50a083c369364440

@MikeRenwick-ICG do you happen to have a fix?
